### DISCUSSION TITLE
Gcode for UBL

### DISCUSSION THREAD

**Discussion by olskar** [Jyers/Marlin#950] [Q&A] [answered]
Hello!

Trying to figure out UBL. First I run below commands to make a mesh, and then have the start g-code below that in cura. 

Am I understanding this right?

```
M140 S60 ; Warm bed to 60C.

M190 S60 ; Wait until bed is warm so we're more accurate.

G28 ; Home XYZ.

G29 P1 ; Do automated probing of the bed.

G29 P3 ; Fill in the rest.

G29 S1 ; Save UBL mesh points to EEPROM in Slot 1.

G29 F 10.0 ; Set Fade Height for correction at 10.0 mm.

G29 A ; Activate the UBL System.
```

Then starting g-code could be this?
```
; Ender 3 Custom Start G-code

M140 S{material_bed_temperature_layer_0} ; Set Heat Bed temperature

M190 S{material_bed_temperature_layer_0} ; Wait for Heat Bed temperature

M104 S160; start warming extruder to 160

G28 ; Home all axes

G29 A ; Activate the UBL System.

G29 L1 ; Load UBL

G29 J ; Quick 3-point level

G29 F10.0 ; Fade to 10mm

G92 E0 ; Reset Extruder

M104 S{material_print_temperature_layer_0} ; Set Extruder temperature

G1 X0.1 Y20 Z0.3 F5000.0 ; Move to start position

M109 S{material_print_temperature_layer_0} ; Wait for Extruder temperature

; G1 Z2.0 F3000 ; Move Z Axis up little to prevent scratching of Heat Bed

G1 X0.1 Y200.0 Z0.3 F1500.0 E15 ; Draw the first line

G1 X0.4 Y200.0 Z0.3 F5000.0 ; Move to side a little

G1 X0.4 Y20 Z0.3 F1500.0 E30 ; Draw the second line

G92 E0 ; Reset Extruder

G1 Z2.0 F3000 ; Move Z Axis up little to prevent scratching of Heat Bed

; End of custom start GCode
```

**Comment by gothcha** [comment, marked as answer]
Wow it does not have to be this complex. Simply create the mesh from the levelling menu and store it. View it to check it looks OK. 

Your start code is then fine, but I use G29 L0 as the default slot is 0. Set/check this in the levelling settings.

**Reply by olskar** [reply]
Do I also need M420 S1 @gothcha?

**Reply by gothcha** [reply]
Yes this enables levelling, G29 A does the same thing so use either

**Comment by smooreace** [comment]
So many levels of wow!

There is no absolutely zero logic in your heating process either.

Heat the bed to predefined number
Test bed temp
repeat until target is reached.... THEN start the hot end heating till it
reaches 160

Its unlikely that you will always want a nozzle temp of 160... so pull that
temp from the jobs g-code
Why heat the 2 separately?
Why cycle heating and testing for the bed and not care about the nozzle?

*M140 S60 ; Warm bed to 60C.*

*M190 S60 ; Wait until bed is warm so we're more accurate.*

*G28 ; Home XYZ.*

*G29 P1 ; Do automated probing of the bed.*

*G29 P3 ; Fill in the rest.*

*G29 S1 ; Save UBL mesh points to EEPROM in Slot 1.*

*G29 F 10.0 ; Set Fade Height for correction at 10.0 mm.*

*G29 A ; Activate the UBL System.`*


*Then starting g-code could be this?`; Ender 3 Custom Start G-code*

*M140 S{material_bed_temperature_layer_0} ; Set Heat Bed temperature*

*M190 S{material_bed_temperature_layer_0} ; Wait for Heat Bed temperature*

*M104 S160; start warming extruder to 160*

*G28 ; Home all axes*

*G29 A ; Activate the UBL System.*

*G29 L1 ; Load UBL*

*G29 J ; Quick 3-point level*

*G29 F10.0 ; Fade to 10mm*

*G92 E0 ; Reset Extruder*

*M104 S{material_print_temperature_layer_0} ; Set Extruder temperature*

*G1 X0.1 Y20 Z0.3 F5000.0 ; Move to start position*

*M109 S{material_print_temperature_layer_0} ; Wait for Extruder temperature*

*; G1 Z2.0 F3000 ; Move Z Axis up little to prevent scratching of Heat Bed*

*G1 X0.1 Y200.0 Z0.3 F1500.0 E15 ; Draw the first line*

*G1 X0.4 Y200.0 Z0.3 F5000.0 ; Move to side a little*

*G1 X0.4 Y20 Z0.3 F1500.0 E30 ; Draw the second line*

*G92 E0 ; Reset Extruder*

*G1 Z2.0 F3000 ; Move Z Axis up little to prevent scratching of Heat Bed*

*; End of custom start GCode*

All shortened down and more functional to....

; Ender 3 Custom Start G-code

;------------------------------------------

;*** Start Dual Nozzle/Bed Preheating ***

M140 S{material_bed_temperature_layer_0} ; start preheating the bed

M104 S{material_print_temperature_layer_0} ﻿T0 ; start preheating hotend

G28 ; home

M190 S{material_bed_temperature_layer_0} ; heat to Cura Bed setting

M109 S{material_print_temperature_layer_0} ﻿T0 ; heat to Cura Hotend

;*** End Preheating ***

;*-------------------------------------------------*

;*** Load Bed Mesh ***

M420 S1

;*-------------------------------------------------*

;*** Draw a Nozzle Cleaning line on the Left Side of Bed

G92 E0 ; Reset Extruder

G1 Z2.0 F3000 ; Move Z Axis up little

G1 X0.1 Y200.0 Z0.3 F1500.0 E15 ; Draw the first line G1 X0.4 Y200.0 Z0.3
F5000.0 ; Move to side a little G1 X0.4 Y20 Z0.3 F1500.0 E30 ; Draw the
second line G92 E0 ; Reset Extruder G1 Z2.0 F3000 ; Move Z Axis up little
G1 X5 Y20 Z0.3 F5000.0 ; Move over to prevent blob squish

**Reply by olskar** [reply]
Took it from here.. https://www.reddit.com/r/ender3/comments/i9xt6e/stepbystep_guide_for_ender_3_v2_firmware_w/

As I wrote I am not sure what a good start gcode would be, happy for suggestions :)

**Comment by kted** [comment]
This my start g-code (excluding the octolapse stuff)

M104 S160  ; Start warming extruder to 160, to avoid filament drip
M190 S{material_bed_temperature_layer_0} ; Wait for Heat Bed temperature
G28        ; Home all axes
G29 A      ; Activate the UBL System.
G29 L0     ; Load UBL
G29 J2     ; 4-point level
G29 F10.0  ; Fade to 10mm (just in case)
G92 E0     ; Reset Extruder

G1 Z2.0 F3000                   ; Move Z Axis up little to prevent scratching of Heat Bed
G1 X0.1 Y20 Z0.3 F5000.0        ; Move to start position
M109 S{material_print_temperature_layer_0} ; Wait for extruder temperature to reach desire temp
G1 X0.1 Y200.0 Z0.3 F1500.0 E15 ; Draw the first line
G1 X0.4 Y200.0 Z0.3 F5000.0     ; Move to side a little
G1 X0.4 Y20 Z0.3 F1500.0 E30    ; Draw the second line
G92 E0                          ; Reset Extruder
G1 Z2.0 F3000                   ; Move Z Axis up little to prevent scratching of Heat Bed
G1 X5 Y20 Z0.3 F5000.0          ; Move over to prevent blob squish


The idea is to pre-heat the extruder to 160C, to avoid filament drip on the table, while we wait for the table to reach our desired temperature.
Then auto-tilt the bed, home the printhead, and wait for the nozzle to reach the desired material temperature. Any filament oozing now, will be taken care of while printing the purge line.

I print almost exclusively remotely, so this is a life saver, since usually there is no one arround  to pick any filament hanging from the nozzle before starting the print.
